lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -254,43 +254,49 @@ public void run() {
 	 * transport events/notifications
 	 ***************************************/
 
-	void onMessage(ProtocolMessage message) {
-		if(protocolListener != null)
-			protocolListener.onRawMessage(message);
-		switch(message.action) {
-		case HEARTBEAT:
-			onHeartbeat(message);
-			break;
-		case ERROR:
-			ErrorInfo reason = message.error;
-			if(reason == null)
-				Log.e(TAG, "onMessage(): ERROR message received (no error detail)");
-			else
-				Log.e(TAG, "onMessage(): ERROR message received; message = " + reason.message + "; code = " + reason.code);
+	void onMessage(ProtocolMessage message) throws AblyException {
```

Comment:
I think this is a detail, but I'm not sure I want these exceptions to propagate back to the caller here.

There's a catch around the call in `WsClient`, but that catch was intended to be handling any deserialisation failure of the message. An error in processing a correctly deserialised message, I think, should be handled here.

You might make this explicit by changing the code in `WsClient` to perform only the decoding within the try/catch, remove the throws declaration here, and handle the exception in the catch at the bottom of this method.